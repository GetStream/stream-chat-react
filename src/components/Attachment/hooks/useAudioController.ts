import { useCallback, useEffect, useRef, useState } from 'react';

export const PROGRESS_UPDATE_INTERVAL = 100;

export const useAudioController = () => {
  const [isPlaying, setIsPlaying] = useState(false);
  const [progress, setProgress] = useState(0);

  const audioRef = useRef<HTMLAudioElement | null>(null);

  const togglePlay = useCallback(() => {
    setIsPlaying((playing) => !playing);
  }, []);

  const seek = useCallback<React.MouseEventHandler<HTMLDivElement>>(
    ({ clientX, currentTarget }) => {
      if (!audioRef.current) return;

      const { width, x } = currentTarget.getBoundingClientRect();

      const ratio = (clientX - x) / width;

      if (!isPlaying) setProgress(ratio * 100);

      audioRef.current.currentTime = ratio * audioRef.current.duration;
    },
    [isPlaying],
  );

  useEffect(() => {
    if (!audioRef.current || !isPlaying) return;

    const interval = window.setInterval(() => {
      if (!audioRef.current) return;

      const { currentTime, duration } = audioRef.current;

      setProgress((currentTime / duration) * 100);

      if (currentTime === duration) setIsPlaying(false);
    }, PROGRESS_UPDATE_INTERVAL);

    audioRef.current.play();

    return () => {
      audioRef.current?.pause();

      window.clearInterval(interval);
    };
  }, [isPlaying]);

  return {
    audioRef,
    isPlaying,
    progress,
    seek,
    togglePlay,
  };
};
