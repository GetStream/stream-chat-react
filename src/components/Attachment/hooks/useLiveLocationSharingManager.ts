import { LiveLocationManager } from 'stream-chat';
import { useEffect, useMemo } from 'react';
import type {
  ExtendableGenerics,
  LiveLocationManagerConstructorParameters,
  StreamChat,
} from 'stream-chat';

export const useLiveLocationSharingManager = <SCG extends ExtendableGenerics>({
  client,
  retrieveAndDeserialize,
  serializeAndStore,
  watchLocation,
}: Omit<LiveLocationManagerConstructorParameters<SCG>, 'client'> & {
  client?: StreamChat<SCG> | null;
}) => {
  const manager = useMemo(() => {
    if (!client) return null;

    return new LiveLocationManager<SCG>({
      client,
      retrieveAndDeserialize,
      serializeAndStore,
      watchLocation,
    });
  }, [client, retrieveAndDeserialize, serializeAndStore, watchLocation]);

  useEffect(() => {
    if (!manager) return;

    manager.registerSubscriptions();

    return () => manager.unregisterSubscriptions();
  }, [manager]);

  return manager;
};
