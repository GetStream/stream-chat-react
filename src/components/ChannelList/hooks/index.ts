export * from './useChannelDeletedListener';
export * from './useChannelHiddenListener';
export * from './useChannelTruncatedListener';
export * from './useChannelUpdatedListener';
export * from './useChannelVisibleListener';
export * from './useConnectionRecoveredListener';
export * from './useMessageNewListener';
export * from './useMobileNavigation';
export * from './useNotificationAddedToChannelListener';
export * from './useNotificationMessageNewListener';
export * from './useNotificationRemovedFromChannelListener';
export * from './usePaginatedChannels';
export * from './useUserPresenceChangedListener';
export * from './useChannelMembershipState';
