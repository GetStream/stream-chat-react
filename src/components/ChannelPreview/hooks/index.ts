export { useChannelPreviewInfo } from './useChannelPreviewInfo';
export { MessageDeliveryStatus } from './useMessageDeliveryStatus';
