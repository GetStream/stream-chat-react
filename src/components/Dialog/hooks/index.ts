export * from './useDialog';
export type { PopperLikePlacement } from './usePopoverPosition';
