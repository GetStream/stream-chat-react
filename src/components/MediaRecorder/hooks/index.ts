export type { RecordingController } from './useMediaRecorder';
