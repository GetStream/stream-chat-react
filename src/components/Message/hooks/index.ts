export * from './useActionHandler';
export * from './useDeleteHandler';
export * from './useEditHandler';
export * from './useFlagHandler';
export * from './useMentionsHandler';
export * from './useMuteHandler';
export * from './useOpenThreadHandler';
export * from './usePinHandler';
export * from './useReactionHandler';
export * from './useRetryHandler';
export * from './useUserHandler';
export * from './useUserRole';
