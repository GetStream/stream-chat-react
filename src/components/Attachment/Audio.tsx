import React from 'react';

import type { Attachment } from 'stream-chat';

import { PauseIcon, PlayTriangleIcon } from './icons';

import { FileSizeIndicator } from './FileSizeIndicator';
import { DownloadButton } from './DownloadButton';
import { useAudioController } from './hooks/useAudioController';

import { useChatContext } from '../../context/ChatContext';

import type { DefaultStreamChatGenerics } from '../../types/types';

export type AudioProps<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = {
  og: Attachment<StreamChatGenerics>;
};

const AudioV1 = ({ og }: AudioProps) => {
  const { asset_url, description, image_url, text, title } = og;
  const { audioRef, isPlaying, progress, togglePlay } = useAudioController();

  return (
    <div className='str-chat__audio'>
      <div className='str-chat__audio__wrapper'>
        <audio ref={audioRef}>
          <source data-testid='audio-source' src={asset_url} type='audio/mp3' />
        </audio>
        <div className='str-chat__audio__image'>
          <div className='str-chat__audio__image--overlay'>
            {!isPlaying ? (
              <button
                className='str-chat__audio__image--button'
                data-testid='play-audio'
                onClick={togglePlay}
              >
                <svg height='40' viewBox='0 0 64 64' width='40' xmlns='http://www.w3.org/2000/svg'>
                  <path
                    d='M32 58c14.36 0 26-11.64 26-26S46.36 6 32 6 6 17.64 6 32s11.64 26 26 26zm0 6C14.327 64 0 49.673 0 32 0 14.327 14.327 0 32 0c17.673 0 32 14.327 32 32 0 17.673-14.327 32-32 32zm13.237-28.412L26.135 45.625a3.27 3.27 0 0 1-4.426-1.4 3.319 3.319 0 0 1-.372-1.47L21 23.36c-.032-1.823 1.41-3.327 3.222-3.358a3.263 3.263 0 0 1 1.473.322l19.438 9.36a3.311 3.311 0 0 1 .103 5.905z'
                    fillRule='nonzero'
                  />
                </svg>
              </button>
            ) : (
              <button
                className='str-chat__audio__image--button'
                data-testid='pause-audio'
                onClick={togglePlay}
              >
                <svg height='40' viewBox='0 0 64 64' width='40' xmlns='http://www.w3.org/2000/svg'>
                  <path
                    d='M32 58.215c14.478 0 26.215-11.737 26.215-26.215S46.478 5.785 32 5.785 5.785 17.522 5.785 32 17.522 58.215 32 58.215zM32 64C14.327 64 0 49.673 0 32 0 14.327 14.327 0 32 0c17.673 0 32 14.327 32 32 0 17.673-14.327 32-32 32zm-7.412-45.56h2.892a2.17 2.17 0 0 1 2.17 2.17v23.865a2.17 2.17 0 0 1-2.17 2.17h-2.892a2.17 2.17 0 0 1-2.17-2.17V20.61a2.17 2.17 0 0 1 2.17-2.17zm12.293 0h2.893a2.17 2.17 0 0 1 2.17 2.17v23.865a2.17 2.17 0 0 1-2.17 2.17h-2.893a2.17 2.17 0 0 1-2.17-2.17V20.61a2.17 2.17 0 0 1 2.17-2.17z'
                    fillRule='nonzero'
                  />
                </svg>
              </button>
            )}
          </div>
          {image_url && <img alt={`${description}`} src={image_url} />}
        </div>
        <div className='str-chat__audio__content'>
          <span className='str-chat__audio__content--title'>
            <strong>{title}</strong>
          </span>
          <span className='str-chat__audio__content--subtitle'>{text}</span>
          <div className='str-chat__audio__content--progress'>
            <div
              data-progress={progress}
              data-testid='audio-progress'
              role='progressbar'
              style={{ width: `${progress}%` }}
            />
          </div>
        </div>
      </div>
    </div>
  );
};

type PlayButtonProps = {
  isPlaying: boolean;
  onClick: () => void;
};

export const PlayButton = ({ isPlaying, onClick }: PlayButtonProps) => (
  <button
    className='str-chat__message-attachment-audio-widget--play-button'
    data-testid={isPlaying ? 'pause-audio' : 'play-audio'}
    onClick={onClick}
  >
    {isPlaying ? <PauseIcon /> : <PlayTriangleIcon />}
  </button>
);

type ProgressBarProps = {
  progress: number;
} & Pick<React.ComponentProps<'div'>, 'onClick'>;

export const ProgressBar = ({ onClick, progress }: ProgressBarProps) => (
  <div
    className='str-chat__message-attachment-audio-widget--progress-track'
    data-progress={progress}
    data-testid='audio-progress'
    onClick={onClick}
    role='progressbar'
    style={{
      background: `linear-gradient(
		 to right, 
		 var(--str-chat__primary-color),
		 var(--str-chat__primary-color) ${progress}%,
		 var(--str-chat__disabled-color) ${progress}%,
		 var(--str-chat__disabled-color)
	  )`,
    }}
  >
    <div
      className='str-chat__message-attachment-audio-widget--progress-slider'
      style={{ left: `${progress}px` }}
    />
  </div>
);

const AudioV2 = ({ og }: AudioProps) => {
  const { asset_url, file_size, title } = og;
  const { audioRef, isPlaying, progress, seek, togglePlay } = useAudioController();

  if (!asset_url) return null;

  const dataTestId = 'audio-widget';
  const rootClassName = 'str-chat__message-attachment-audio-widget';

  return (
    <div className={rootClassName} data-testid={dataTestId}>
      <audio ref={audioRef}>
        <source data-testid='audio-source' src={asset_url} type='audio/mp3' />
      </audio>
      <div className='str-chat__message-attachment-audio-widget--play-controls'>
        <PlayButton isPlaying={isPlaying} onClick={togglePlay} />
      </div>
      <div className='str-chat__message-attachment-audio-widget--text'>
        <div className='str-chat__message-attachment-audio-widget--text-first-row'>
          <div className='str-chat__message-attachment-audio-widget--title'>{title}</div>
          <DownloadButton assetUrl={asset_url} />
        </div>
        <div className='str-chat__message-attachment-audio-widget--text-second-row'>
          <FileSizeIndicator fileSize={file_size} />
          <ProgressBar onClick={seek} progress={progress} />
        </div>
      </div>
    </div>
  );
};

const UnMemoizedAudio = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>(
  props: AudioProps<StreamChatGenerics>,
) => {
  const { themeVersion } = useChatContext<StreamChatGenerics>('Audio');

  return themeVersion === '1' ? <AudioV1 {...props} /> : <AudioV2 {...props} />;
};

/**
 * Audio attachment with play/pause button and progress bar
 */
export const Audio = React.memo(UnMemoizedAudio) as typeof UnMemoizedAudio;
