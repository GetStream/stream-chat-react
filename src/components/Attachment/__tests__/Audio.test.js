import React from 'react';
import prettybytes from 'pretty-bytes';
import { cleanup, fireEvent, render, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import { Audio } from '../Audio';
import { PROGRESS_UPDATE_INTERVAL } from '../hooks/useAudioController';

import { ChatContext } from '../../../context/ChatContext';

import { generateAudioAttachment } from 'mock-builders';

const AUDIO = generateAudioAttachment();

const renderComponent = (
  props = {
    chatContext: { themeVersion: '1' },
    og: AUDIO,
  },
) =>
  render(
    <ChatContext.Provider value={props.chatContext}>
      <Audio og={props.og} />
    </ChatContext.Provider>,
  );

const playButtonTestId = 'play-audio';
const pauseButtonTestId = 'pause-audio';

describe('Audio', () => {
  beforeAll(() => {
    // jsdom doesn't define these, so mock them instead
    // see https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement#Methods
    jest.spyOn(HTMLMediaElement.prototype, 'play').mockImplementation(() => {});
    jest.spyOn(HTMLMediaElement.prototype, 'pause').mockImplementation(() => {});
  });
  afterEach(cleanup);

  it('in v1 should render title and render the image with description as alt tag', () => {
    const { container } = renderComponent({
      chatContext: { themeVersion: '1' },
      og: { ...AUDIO, title: 'deterministic' },
    });

    expect(container).toMatchSnapshot();
  });

  it('in v2 uploaded should render title and file size', () => {
    const { container, getByText } = renderComponent({
      chatContext: { themeVersion: '2' },
      og: AUDIO,
    });

    expect(getByText(AUDIO.title)).toBeInTheDocument();
    expect(getByText(prettybytes(AUDIO.file_size))).toBeInTheDocument();
    expect(container.querySelector('img')).not.toBeInTheDocument();
  });

  it('in v2 should show the correct progress after clicking to the middle of a progress bar (seeking)', async () => {
    const { getByTestId } = renderComponent({ chatContext: { themeVersion: '2' }, og: AUDIO });

    jest
      .spyOn(HTMLDivElement.prototype, 'getBoundingClientRect')
      .mockImplementationOnce(() => ({ width: 120, x: 0 }));

    jest.spyOn(HTMLAudioElement.prototype, 'currentTime', 'set').mockImplementationOnce(() => {});

    fireEvent.click(getByTestId('audio-progress'), {
      clientX: 60,
    });

    await waitFor(() => {
      expect(getByTestId('audio-progress')).toHaveAttribute('data-progress', '50');
    });
  });

  describe.each([['1'], ['2']])('version %s', (themeVersion) => {
    it('should render an audio element with the right source', () => {
      const { getByTestId } = renderComponent({ chatContext: { themeVersion }, og: AUDIO });

      const source = getByTestId('audio-source');

      expect(source).toBeInTheDocument();
      expect(source.src).toBe(AUDIO.asset_url);
      expect(source.parentElement).toBeInstanceOf(HTMLAudioElement);
    });

    it('should show the correct button if the song is paused/playing', async () => {
      const { queryByTestId } = renderComponent({ chatContext: { themeVersion }, og: AUDIO });

      const playButton = () => queryByTestId(playButtonTestId);
      const pauseButton = () => queryByTestId(pauseButtonTestId);

      expect(await playButton()).toBeInTheDocument();
      expect(await pauseButton()).not.toBeInTheDocument();

      fireEvent.click(playButton());

      expect(await playButton()).not.toBeInTheDocument();
      expect(await pauseButton()).toBeInTheDocument();

      fireEvent.click(pauseButton());

      expect(await playButton()).toBeInTheDocument();
      expect(await pauseButton()).not.toBeInTheDocument();
    });

    it(`should poll for progress every ${PROGRESS_UPDATE_INTERVAL}ms if the file is played, and stop doing that when it is paused`, () => {
      const { getByTestId } = renderComponent({ chatContext: { themeVersion }, og: AUDIO });

      let intervalId;
      const setIntervalSpy = jest.spyOn(window, 'setInterval').mockImplementationOnce(() => {
        intervalId = 'something';
        return intervalId;
      });
      const clearIntervalSpy = jest.spyOn(window, 'clearInterval');

      fireEvent.click(getByTestId(playButtonTestId));
      expect(setIntervalSpy).toHaveBeenCalledWith(expect.any(Function), PROGRESS_UPDATE_INTERVAL);

      fireEvent.click(getByTestId(pauseButtonTestId));
      expect(clearIntervalSpy).toHaveBeenCalledWith(intervalId);
    });

    it('should clean up the progress interval if the component is unmounted while the file is playing', () => {
      const { getByTestId, unmount = cleanup } = renderComponent({
        chatContext: { themeVersion },
        og: AUDIO,
      });

      let intervalId;
      jest.spyOn(window, 'setInterval').mockImplementationOnce(() => {
        intervalId = 'something';
        return intervalId;
      });
      fireEvent.click(getByTestId(playButtonTestId));

      const clearIntervalSpy = jest.spyOn(window, 'clearInterval');

      unmount();

      expect(clearIntervalSpy).toHaveBeenCalledWith(intervalId);
    });

    it('should show the correct progress', async () => {
      const { getByTestId } = renderComponent({ chatContext: { themeVersion }, og: AUDIO });

      jest.spyOn(HTMLAudioElement.prototype, 'duration', 'get').mockImplementationOnce(() => 100);
      jest.spyOn(HTMLAudioElement.prototype, 'currentTime', 'get').mockImplementationOnce(() => 50);

      fireEvent.click(getByTestId(playButtonTestId));

      await waitFor(() => {
        expect(getByTestId('audio-progress')).toHaveAttribute('data-progress', '50');
      });
    });
  });
});
