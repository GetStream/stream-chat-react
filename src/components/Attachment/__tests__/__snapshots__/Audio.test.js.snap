// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Audio in v1 should render title and render the image with description as alt tag 1`] = `
<div>
  <div
    class="str-chat__audio"
  >
    <div
      class="str-chat__audio__wrapper"
    >
      <audio>
        <source
          data-testid="audio-source"
          src="http://www.jackblack.com/tribute.mp3"
          type="audio/mp3"
        />
      </audio>
      <div
        class="str-chat__audio__image"
      >
        <div
          class="str-chat__audio__image--overlay"
        >
          <button
            class="str-chat__audio__image--button"
            data-testid="play-audio"
          >
            <svg
              height="40"
              viewBox="0 0 64 64"
              width="40"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M32 58c14.36 0 26-11.64 26-26S46.36 6 32 6 6 17.64 6 32s11.64 26 26 26zm0 6C14.327 64 0 49.673 0 32 0 14.327 14.327 0 32 0c17.673 0 32 14.327 32 32 0 17.673-14.327 32-32 32zm13.237-28.412L26.135 45.625a3.27 3.27 0 0 1-4.426-1.4 3.319 3.319 0 0 1-.372-1.47L21 23.36c-.032-1.823 1.41-3.327 3.222-3.358a3.263 3.263 0 0 1 1.473.322l19.438 9.36a3.311 3.311 0 0 1 .103 5.905z"
                fill-rule="nonzero"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="str-chat__audio__content"
      >
        <span
          class="str-chat__audio__content--title"
        >
          <strong>
            deterministic
          </strong>
        </span>
        <span
          class="str-chat__audio__content--subtitle"
        />
        <div
          class="str-chat__audio__content--progress"
        >
          <div
            data-progress="0"
            data-testid="audio-progress"
            role="progressbar"
            style="width: 0%;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
